### PR TITLE
End iterator on GC or db close

### PR DESCRIPTION
**Pull request by vweevers**
Removing the need for the user to do `iterator.end()`, following the suggestion by @chjj in https://github.com/Level/leveldown/pull/597#issuecomment-472756492.

For early review; needs a CI run, some light refactoring and comments to clarify the order of things.

### CODE REVIEW

**Comment by vweevers** [comment]
office_so-close

File: test/iterator-test.js [hunk 1]
```diff
@@ -66,23 +66,17 @@ make('iterator optimized for seek', function (db, t, done) {
...
-    if (cnt++ === 0) {
-      t.ifError(err, 'no error from next()')
-    } else {
-      t.equal(err.message, 'iterator has ended')
-      hadError = true
-    }
-    if (key !== undefined) { ite.next(loop) }
+    nexted = true
+    t.ifError(err, 'no error from next()')
+    // if (key !== undefined) { ite.next(loop) }
```
Comment (vweevers) [inline test/iterator-test.js]: Gotta restore the previous behavior here (calling `it.next()` after `db.close()` should throw).

**Comment by vweevers** [comment]
happy cage

File: binding.cc [hunk 2]
```diff
@@ -565,24 +574,53 @@ struct Iterator {
...
+      database_->DetachIterator(id_);
+    }
   }
 
-  leveldb::Status IteratorStatus () {
-    return dbIterator_->status();
+  void DeleteReference (napi_env env) {
+    // TODO: we seem to be hitting nodejs/node#24494 or a related bug, in node < 11.
+    // When uncommented we get a segfault in the finalizer caller.
+    // napi_delete_reference(env, ref_);
```
Comment (vweevers) [inline binding.cc]: I don't know how to fix this. We need to delete the reference to avoid leaking, but calling `napi_delete_reference` in the finalizer callback results in a segfault (in node 8 and 10). Might be https://github.com/nodejs/node/pull/24494?

**Comment by vweevers** [comment]
Gonna take a break from this for a few days. I'm close and the code is fairly clean now, but I feel like I'm stuck in a bike shed and need a breather to regain clarity of the big picture.

File: binding.cc [hunk 3]
```diff
@@ -836,13 +891,32 @@ struct CloseWorker final : public BaseWorker {
...
+        iterator->IncrementReferenceCount();
+        iterator->closing_ = true;
+      });
+    }
 
   ~CloseWorker () {}
 
   void DoExecute () override {
+    database_->VisitIterators([&] (Iterator* iterator) {
+      iterator->Close();
```
Comment (vweevers) [inline binding.cc]: Hmm, GC-ed iterators are closed on the main thread so might as well do the same\* here. For this PR to be acceptable, we need closing to be fast anyway (AFAIK there's no disk access involved in e.g. releasing the snapshot so it might be fine).

\* I.e. move the `iterator->Close()` call up, which means I can then also replace usage of `closing_` with `closed_`, move the `Detach()` call back into `Close()`, and forego the reference counting (possibly even altogether, because it is not strictly necessary while nexting, which is already guarded from GC by the fact that `next()` callback in JS land has a reference to the iterator). Jeez :)

**Comment by chjj** [comment]
Wow, great work! I've thought about this for a long time but never realized the complexity involved in actually implementing it. I'll try to play around with this and see if I can break it.

**Comment by vweevers** [comment]
@chjj Don't dive too deep yet because I plan to simplify it (but haven't had much time)

File: binding.cc (hunk 3 above)
Comment (vweevers) [reply binding.cc]: Reconsidered. On database close, there may be multiple iterators left (with references in userland) so I prefer to do the work in a worker pool thread. And I was wrong about the iterator always being persistent while nexting (there's a window in between `NextWorker::DoComplete` and `::DoFinally` where it can be GC-ed, I added a test for this edge case).

**Comment by vweevers** [comment]
Flowchart:

leveldown iterator flowchart

**Comment by vweevers** [comment]
Gonna open it up for review (and @chjj go! kick the tires!) though there's some cleanup left (reverting the repeated tests and then squashing). I prefer to keep repeating tests when changes are requested.

File: binding.cc [hunk 4]
```diff
@@ -1412,87 +1455,22 @@ NAPI_METHOD(iterator_seek) {
     }
   }
 
-  NAPI_RETURN_UNDEFINED();
-}
-
-/**
- * Worker class for ending an iterator
- */
-struct EndWorker final : public BaseWorker {
```
Comment (peakji) [inline binding.cc]: I like the idea of ending on GC, but it will be nice if we keep the ability to manually end an iterator as well.

One may not want to fully rely on GC on a busy server if there is a clean way to eagerly release an iterator (and the underlying snapshot and resources).

File: binding.cc [hunk 5]
```diff
@@ -1827,7 +1800,6 @@ NAPI_INIT() {
 
   NAPI_EXPORT_FUNCTION(iterator_init);
   NAPI_EXPORT_FUNCTION(iterator_seek);
-  NAPI_EXPORT_FUNCTION(iterator_end);
```
Comment (peakji) [inline binding.cc]: On the perspective of the level-ecosystem, we cannot expect iterators of all kinds of backends to end on GC automatically without a problem, it is implementation specific. Maybe later we should mark `iterator.end` as optional in `abstract-leveldown`?

Comment (vweevers) [reply binding.cc]: Nothing has to change in `abstract-leveldown` for now. I first want to explore the idea in `leveldown`.

File: iterator.js [hunk 6]
```diff
@@ -54,8 +54,9 @@ Iterator.prototype._next = function (callback) {
 }
 
 Iterator.prototype._end = function (callback) {
+  // TODO: is any resource in C++ land tied to this, or can we skip this?
   delete this.cache
```
[the remark this answers is not in the document]
Comment (vweevers) [reply iterator.js]: Right

File: binding.cc (hunk 4 above)
Comment (vweevers) [reply binding.cc]: I prefer either/or. Explicit resource management, or GC. Supporting both is not impossible (the `EndWorker` could do the same for a single iterator as what the `CloseWorker` does for all iterators now), but it raises the question of what we recommend as best practice - and whether modules that wrap iterators should end explicitly or not.

Would be nice to do some real world benchmarks.

Comment (peakji) [reply binding.cc]: 👍 for benchmarks. I thought ending on GC was purely for better promise support as @chjj mentioned in https://github.com/Level/leveldown/pull/597#issuecomment-472756492.

Comment (vweevers) [reply binding.cc]: I'm not sure _what_ to benchmark. Ending on GC will have some cost, but what cost is acceptable?

Comment (peakji) [reply binding.cc]: One of my private project use iterators for prefix searching, with lots of concurrent `seek` `next` and (early) `end`. I would like to benchmark the new GC-version with my project (especially on write intensive workloads), but I'm pretty busy for the next few days...

Shall we release **v5.0.0** first and work on this PR later?

Comment (vweevers) [reply binding.cc]: Already released it :)

Comment (peakji) [reply binding.cc]: My main concern is about the underlying snapshot. I remember that holding snapshots might affect compression of new keys after the ** SequenceNumber**, so LevelDB suggests to release snapshots as soon as possible.

Comment (vweevers) [reply binding.cc]: I'll assign you to this PR, no rush.

**Comment by peakji** [comment]
🤔 In order to benchmark the two implementations, I've routed some real world workload to a test server. I started with the original [v5.0.0](https://github.com/Level/leveldown/releases/tag/v5.0.0), but it crashed on segmentation fault (**status=11/SEGV**) after about an hour.

Now I'm trying to reproduce the issue in local environment, I think it should be top priority since v5.0.0+ is already on NPM. I'll open a new issue or PR after locating the cause, and go back to this PR later.

**Comment by vweevers** [comment]
> I think it should be top priority since v5.0.0+ is already on NPM.

👍

**Comment by vweevers** [comment]
@peakji Can you share some details about the test you did? What kind of operations, open/close states, a stack trace if you have it, etc?

**Comment by peakji** [comment]
@vweevers It was a segment fault so no useful stack trace. I've enabled core dump yesterday but unfortunately it didn't break (till now), and I have a feeling that it will break tonight 😜. I'll do a back trace with `lldb` and share the result after I got the core dump file.

In addition, I can confirm that the issue is not related to `Iterator`. The test software has two repeating phases: upsert and iterate, and it broke during a upsert phase, so we can focus on checking `get`, `put`, `del`, and `Batch`. I guess we can start with looking for dereferencing NULL?

**Comment by vweevers** [comment]
> it broke during a upsert phase, so we can focus on checking `get`, `put`, `del`, and `Batch`. I guess we can start with looking for dereferencing NULL?

Does the software open and close the database during this phase? If not, that rules out an obvious suspect

**Comment by peakji** [comment]
> Does the software open and close the database during this phase? If not, that rules out an obvious suspect

No, the software opens a database on start and uses it forever, and I'm sure it opened successfully.

**Comment by peakji** [comment]
The software is still up and running, no sign of segment fault, monitor info looks perfect 🙄. I know its kinda weird to wish your own software to crash, but nothing could be fixed by simply enabling core dump and reboot...

In fact what worries me this most is that the segfault is NOT related to `Iterator`, because it means that:

1. We cannot expect this PR to fix the issue.

2. If we [share binding.cc -> rocksdb](https://github.com/Level/rocksdb/issues/100) or others, the issue might come along as well.

@vweevers With your knowledge of the current code base, if you could review and confirm there is no obvious illegal memory access (e.g. out-of-bound, dereferencing NULL, ...), then I'd rather suspect the segfault was caused by the old `node v10.4.1` on the server. Sorry that I can't be of any useful help without a dump file, still haven't got the time to learn N-API.

**Comment by vweevers** [comment]
Will hunt for issues at the end of this week. Do your operations have a predefined order (e.g. a get followed by a del) or is it random? Are you using `db.batch([])` (array form) or `db.batch()` (chained form)?

I will deploy `leveldown@5` to a production system myself in the next 2 weeks, but that one only does `db.batch([])` (array form) and `db.iterator()` operations.

**Comment by peakji** [comment]
> Do your operations have a predefined order (e.g. a get followed by a del) or is it random?

Yes, its always `get` -> `put` and/or `del` -> `batch`(s).  However, I have two queues running simultaneously, so it's actually random from the database perspective. The workflow is simple: first `get` some metadata, then update with `put` and/or remove stale metadata with `del`, then finally write data chunks using one or more `batch`s.

> Are you using db.batch([]) (array form) or db.batch() (chained form)?

I am using the chained form, with mixed `put` and `del` operations. The average number of keys per `batch` is quite big (~hundreds of `put`s and a couple of `del`s), since my original intention was to benchmark the two `Iterator` implementations.

> I will deploy leveldown@5 to a production system myself in the next 2 weeks, but that one only does db.batch([]) (array form) and db.iterator() operations.

I will keep core dump enabled for the next ... umm... forever! until it breaks!

**Comment by vweevers** [comment]
Thanks! I made an edit to my questions just a minute ago before you answered:

> Lastly, are you reading/writing buffer or string keys/values?

**Comment by peakji** [comment]
> Lastly, are you reading/writing buffer or string keys/values?

I'm using buffers for both key and value in all the operations.

**Comment by vweevers** [comment]
I personally need #32 more than I need this (in fact I don't need this at all).

@peakji Any objections to me doing #32 first (utilizing `PriorityWorker` for all operations), releasing that as semver-patch, and us getting back to this PR later?

**Comment by peakji** [comment]
> releasing that as semver-patch, and us getting back to this PR later?

👍 I'll deploy the new patched version on a separated server.

**Comment by vweevers** [comment]
Released 5.0.2

I'll rebase this branch later.

**Comment by vweevers** [comment]
Rebased. I'll also see if I can move some of the smaller refactorings I did here into separate PRs.

**Comment by vweevers** [comment]
@peakji Good news (or bad, depending how you look at it): I found a segfault. Happens when a chained batch is GC-ed in between `batch.write(callback)` and `BatchWriteWorker->DoExecute()`.

**Comment by peakji** [comment]
@vweevers Absolutely good news! 

This solved my doubts, in fact I thought it might be related to `batch`, but not about GC. I was creating a lot of batches in my code, but not all of them were being populated. I thought it was the empty batches which caused the memory error.

BTW, what do you think is the best practice for empty batches (created but unused)? Call `write` or just leave them to the collector?

**Comment by vweevers** [comment]
> BTW, what do you think is the best practice for empty batches (created but unused)? Call write or just leave them to the collector?

There's no benefit to calling `write()`, I think. Especially when #619 lands, which makes `BatchWriteWorker->DoExecute()` a noop on empty batches. And when #621 lands, calling write would merely delay GC (and unnecessarily acquire a thread in the worker pool thread).

**Comment by peakji** [comment]
Just reviewed the new PRs, nice work @vweevers ! These should make a new semver-patch?

I'm now working on a standalone benchmark script for this PR (finally!).

**Comment by vweevers** [comment]
> These should make a new semver-patch?

Yup! Will try to get that out today or tomorrow.

> I'm now working on a standalone benchmark script for this PR (finally!).

Awesome!

**Comment by vweevers** [comment]
(Be advised, GitHub shows the commits in the wrong order due to rebasing)

**Comment by vweevers** [comment]
And gaah, I have to rebase _again_ 😄

**Comment by vweevers** [comment]
@peakji @chjj @ralphtheninja A summary so we're all on the same page:

- Ending iterators on GC degrades performance, because holding too many snapshots negatively affects compaction of new keys (#627)
- Ending iterators automatically in `for await..of`, even when an error is thrown by the user in the body of the loop, is feasible (https://github.com/Level/abstract-leveldown/pull/338)
- That invalidates the original reason for exploring the idea of ending on GC.

Unless there are other reasons for wanting to end on GC, I propose we close this.

One possible reason is "safety" - i.e. when the user forgets to call `end()`. IMO that's hand-holding, not worth the added code complexity.

**Comment by peakji** [comment]
> One possible reason is "safety" - i.e. when the user forgets to call end(). IMO that's hand-holding, not worth the added code complexity.

👍 not worth the added code complexity.

Someone forgetting to call `end()` on iterators may also forget to call `end()` on http requests, file streams, database connections, sockets ... 😛
